### ISSUE TITLE
Upgrade issue with SQL migrations on multi store

### ISSUE THREAD

**Issue by Seb33300** [PrestaShop/PrestaShop#25303]
#### Describe the bug

On a multi store install, the `PS_VERSION_DB` is set in the `ps_configuration` table with `NULL` in the `id_shop` column.

When upgrading Prestashop using the manual upgrade script in `install/upgrade/upgrade.php`, SQL migrations are properly processed, but a new line in `ps_configuration` is created for `PS_VERSION_DB` with `id_shop = 1` to store the upgraded version number.
The previous line with NULL in `id_shop` is not updated and keep the value of the old version.

On the next upgrade, Prestashop read the value from the line with `NULL` in the `id_shop` and try to re run SQL migrations again which cause SQL error.

#### Expected behavior

After upgrade, Prestashop should update the existing `PS_VERSION_DB` and not create a new line.

#### Steps to Reproduce

1. Install Prestashop 1.7.7.1
1. Enable multi store and create a 2nd store
2. Upgrade to Prestashop 1.7.7.4
3. See 2 lines in `ps_configuration` for `PS_VERSION_DB` key
4. Upgrade to Prestashop 1.7.7.5
5. See that upgrade script is executing SQL migration for 1.7.7.2 again

#### Additional information

* PrestaShop version: 1.7.7.4
* PHP version: 7.2

**Comment by khouloudbelguith** [comment]
Hi @Seb33300,

In step4, I make an upgrade from PS1.7.7.1 to PS1.7.7.5 using the 1-click upgrade v4.12.0, but I don't have two lines.
I attached a screen record
https://drive.google.com/file/d/15VSzDbtWtKN7xCic9kvv9BTmxjKhaF14/view?usp=sharing

Does this issue occur only when you upgrade to PS1.7.7.4?
>updated

I tried again, I made an upgrade from PS1.7.7.1 toPS1.7.7.4, But I don't have two lines in the ps_configuration about the version.
Thanks to check this screen record
https://drive.google.com/file/d/171NSFvkjCb3kFFLYRHspAHxWpJ4mGFT0/view?usp=sharing

Thanks!

**Comment by Seb33300** [comment]
Hello @khouloudbelguith, @hibatallahAouadni

@khouloudbelguith Sorry, my explanations may not be clear, but when I said `using the upgrade script in install/upgrade/upgrade.php`, I was not talking about the 1-click upgrade module but using [the cli upgrade script](https://devdocs.prestashop.com/1.7/basics/keeping-up-to-date/upgrade/#database-upgrade) during the manual upgrade process: https://devdocs.prestashop.com/1.7/basics/keeping-up-to-date/upgrade/

@hibatallahAouadni As I said in my PR, this is not a fix for this issue, it's just related

**Comment by Seb33300** [comment]
Screenshot on step 4:

![image](https://user-images.githubusercontent.com/915273/125154557-b398cb80-e184-11eb-92ad-31b56c444871.png)

Here is a screenshot of the step 5
As you can see, 1.7.7.2 migration script is run again (it should not) and ends on a `Duplicate entry 'displayAdminListBefore' for key 'alias'` error.

![image](https://user-images.githubusercontent.com/915273/125154419-28b7d100-e184-11eb-9dd7-e78baf26a455.png)

**Comment by Seb33300** [comment]
I finally found where is the bug and created a PR to fix this issue: https://github.com/PrestaShop/PrestaShop/pull/25311

**Comment by hibatallahAouadni** [comment]
Hello @Seb33300 

Could you please check the attached screenshots below with manual upgrade:

On console:

```
PHP Warning: PHP Startup: Unable to load dynamic library 'apcu.so' (tried: /usr/lib/php/20170718/apcu.so (/usr/lib/php/20170718/apcu.so: cannot open shared object file: No such file or directory), /usr/lib/php/20170718/apcu.so.so (/usr/lib/php/20170718/apcu.so.so: cannot open shared object file: No such file or directory)) in Unknown on line 0
<?xml version="1.0" encoding="UTF-8"?><action result="ok" id="">
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : SET SESSION sql_mode='']]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : SET NAMES 'utf8']]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : INSERT IGNORE INTO ps_hook (id_hook, name, title, description, position) VALUES
(NULL, 'displayAdminGridTableBefore', 'Display before Grid table', 'This hook adds new blocks before Grid component table.', '1'),
(NULL, 'displayAdminGridTableAfter', 'Display after Grid table', 'This hook adds new blocks after Grid component table.', '1')]]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : UPDATE ps_hook_module AS hm
INNER JOIN ps_hook AS hfrom ON hm.id_hook = hfrom.id_hook AND hfrom.name = 'displayAdminListBefore'
INNER JOIN ps_hook AS hto ON hto.name = 'displayAdminGridTableBefore'
SET hm.id_hook = hto.id_hook]]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : DELETE FROM ps_hook WHERE name = 'displayAdminListBefore']]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : UPDATE ps_hook_module AS hm
INNER JOIN ps_hook AS hfrom ON hm.id_hook = hfrom.id_hook AND hfrom.name = 'displayAdminListAfter'
INNER JOIN ps_hook AS hto ON hto.name = 'displayAdminGridTableAfter'
SET hm.id_hook = hto.id_hook]]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : DELETE FROM ps_hook WHERE name = 'displayAdminListAfter']]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : INSERT INTO ps_hook_alias (name, alias) VALUES
('displayAdminGridTableBefore', 'displayAdminListBefore'),
('displayAdminGridTableAfter', 'displayAdminListAfter')]]></action>
```

As you can see I didn't manage to reproduce the isssue mentioned in your comment below:

> 1.7.7.2 migration script is run again (it should not) and ends on a Duplicate entry 'displayAdminListBefore' for key 'alias' error.

In the other hand, @florine2623 encounter this error when she try to reach the address `http://<shop_domain>/install/upgrade/upgrade.php`

![image](https://user-images.githubusercontent.com/23308427/125424566-2bc11e88-cc70-44ad-9dff-1a8017fdabfc.png)

---

> On a multi store install, the PS_VERSION_DB is set in the ps_configuration table with NULL in the id_shop column.

About this bug, we need to confirm if this is the expected behavior or not:

**1771 -> 1774:**

![image](https://user-images.githubusercontent.com/23308427/125422913-1a761c66-1d91-4a99-bb5d-7c26f054332b.png)

**1771 -> 1774 -> 1775:**

![image](https://user-images.githubusercontent.com/23308427/125422930-da4f04da-32c6-4f32-a03a-e766d1e5c7cc.png)

Ping @PrestaShop/prestashop-core-developers is this the expected behavior when we manually upgrade our shop?

**Comment by Seb33300** [comment]
If you can see all the `<action result="info" id="1.7.7.2">` lines in console when doing the `1771 -> 1774` and `1774 -> 1775` upgrade, this means SQL migrations are run during both upgrades.
They should be run only 1 time

**Comment by hibatallahAouadni** [comment]
Ping @atomiix wdyt? these DB upgrade have been executing twice (for each upgrade):
```
PHP Warning: PHP Startup: Unable to load dynamic library 'apcu.so' (tried: /usr/lib/php/20170718/apcu.so (/usr/lib/php/20170718/apcu.so: cannot open shared object file: No such file or directory), /usr/lib/php/20170718/apcu.so.so (/usr/lib/php/20170718/apcu.so.so: cannot open shared object file: No such file or directory)) in Unknown on line 0
<?xml version="1.0" encoding="UTF-8"?><action result="ok" id="">
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : SET SESSION sql_mode='']]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : SET NAMES 'utf8']]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : INSERT IGNORE INTO ps_hook (id_hook, name, title, description, position) VALUES
(NULL, 'displayAdminGridTableBefore', 'Display before Grid table', 'This hook adds new blocks before Grid component table.', '1'),
(NULL, 'displayAdminGridTableAfter', 'Display after Grid table', 'This hook adds new blocks after Grid component table.', '1')]]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : UPDATE ps_hook_module AS hm
INNER JOIN ps_hook AS hfrom ON hm.id_hook = hfrom.id_hook AND hfrom.name = 'displayAdminListBefore'
INNER JOIN ps_hook AS hto ON hto.name = 'displayAdminGridTableBefore'
SET hm.id_hook = hto.id_hook]]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : DELETE FROM ps_hook WHERE name = 'displayAdminListBefore']]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : UPDATE ps_hook_module AS hm
INNER JOIN ps_hook AS hfrom ON hm.id_hook = hfrom.id_hook AND hfrom.name = 'displayAdminListAfter'
INNER JOIN ps_hook AS hto ON hto.name = 'displayAdminGridTableAfter'
SET hm.id_hook = hto.id_hook]]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : DELETE FROM ps_hook WHERE name = 'displayAdminListAfter']]></action>
<action result="info" id="1.7.7.2"><![CDATA[[OK] SQL 1.7.7.2 : INSERT INTO ps_hook_alias (name, alias) VALUES
('displayAdminGridTableBefore', 'displayAdminListBefore'),
('displayAdminGridTableAfter', 'displayAdminListAfter')]]></action>
```

Is this the expected behavior?

**Comment by Seb33300** [comment]
@hibatallahAouadni FYI I created a PR to fix the issue on the weekend and it was already approved (but nor merged yet): https://github.com/PrestaShop/PrestaShop/pull/25311

If you do the same test on this PR, SQL migrations should be run only during the upgrade `1771 -> 1774`

**Comment by hibatallahAouadni** [comment]
Ping @PrestaShop/prestashop-core-developers friendly reminder :pray:

**Comment by atomiix** [comment]
@hibatallahAouadni this is definitely not the expected behavior, the fix provided by @Seb33300 seems good! 👍

**Comment by hibatallahAouadni** [comment]
Thanks @atomiix for your quick reply :pray: changing the status to **`To Do`**

**Comment by Progi1984** [comment]
Fixed in #25311 by @Seb33300
